README.md

```diff
@@ -1,137 +1,326 @@
+straightforward interface to integrate the code generation into cmake builds. The generated library contains a c++
+struct with all specified parameters as nested fields. Additionally, methods are generated, which enable dynamic parameter
+updating, parameter copying, and parameter validation.
+
+## Generated code interface
+The generated code is primarily consists of two major component: 1) a pure data struct `struct Params` that contains values of all parameters and 2) a class `ParamListener` that handles parameter declaration, updating, and validation. The general structure is shown below.
+
+```cpp
+namespace cpp_namespace {
+
+struct Params {
+  int param_name = 3;
+  struct {
+    struct{
+      std::string param_name;
+      // arbitrary nesting depth...
+    } nest2;
+  } nest1;
+  // for detecting if the parameter struct has been updated
+  rclcpp::Time __stamp;
+};
+
+class ParamListener {
+ public:
+  ParamListener(rclcpp::ParameterInterface);
+  ParamListener(rclcpp::Node::SharedPtr node)
+    : ParameterListener(node->get_parameters_interface()) {}
+  ParamListener(rclcpp_lifecycle::LifecycleNode::SharedPtr node)
+    : ParameterListener(node->get_parameters_interface()) {}
+
+  // create a copy of current parameter values
+  Params get_params() const;
+
+  // returns true if parameters have been updated since last time get_params was called
+  bool is_old(Params const& other) const;
+
+  // loop over all parameters: perform validation then update
+  rcl_interfaces::msg::SetParametersResult update(const std::vector<rclcpp::Parameter> &parameters);
+
+  // declare all parameters and throw exception if non-optional value is missing or validation fails
+  void declare_params(const std::shared_ptr<rclcpp::node_interfaces::NodeParametersInterface>& parameters_interface);
+
+ private:
+  Params params_;
+};
+
+} // namespace cpp_namespace
 ```
+The structure of the `Params` struct and the logic for declaring and updating parameters is generated from a YAML file specification.
+
+## Usage
+
+Integrating the generate_parameter_library into a project requires the following items:
+<ol>
+  <li>Create YAML parameter codegen file</li>
+  <li>Add parameter library generation to cmake project</li>
+  <li>Integrate generated struct into project source code</li>
+</ol>
+
+Throughout the instructions, the following minimal sample project will be referred to:
+
+```c++
+#include "rclcpp/rclcpp.hpp"
+// additional includes
+
+class MinimalInterface {
+public:
+    MinimalInterface(rclcpp::Node::SharedPtr node) {
+    // Initialization ...
+    }
+    void update(const std_msgs::msg::String::SharedPtr msg) {
+    // Update state  ...
+    }
+    // Member variables ...
+};
+
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+  node = std::make_shared<rclcpp::Node>("minimal_node");
+  // Initialization ...
+  while(rclcpp::ok()){
+    // do update
+  }
+
+  return 0;
+}
```

Comment:
I don't think you need this. You can just describe each step with some sample code around each step to show where it would go.